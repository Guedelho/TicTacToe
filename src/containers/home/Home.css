.container{
  display: flex;
  height: 100vh;
  padding: 10px;
}

.matrix {
  display: grid;
  margin: auto;
  width: 100%;
  max-width: 600px;
  grid-template-columns: 1fr 1fr 1fr;
  border: thin solid #333;
}

.value{
  position: relative;
  padding-top: 100%;
  border: thin solid #333;
  font-size: 100px;
}

.value::before{
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -40%);
}

.player1::before{
  content: 'X';
}

.player2::before{
  content: 'O ';
}

@media (min-width: 480px) {
  .value{
    font-size: 150px;
  }
}

@media (min-width: 768px) {
  .value{
    font-size: 200px;
  }
}