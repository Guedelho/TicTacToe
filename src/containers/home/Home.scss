$black: #343a40;
$green: #28a745;
$blue: #007bff;
$red: #dc3545;

%transition {
  transition: all 0.6s cubic-bezier(0.572, 0, 0.397, 1) 0s;
}

.Home {
  @extend %transition;
  &.blue {
    &.winner{
      background: $blue;
    }
    .matrix,
    .value {
      border-color: $blue;
    }
  }
  &.green {
    &.winner{
      background: $green;
    }
    .matrix,
    .value {
      border-color: $green;
    }
  }
}

.modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9;
  background:#fff;
  &:first-child{
    z-index: 10;
  }
  .content {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 100%;
    padding: 20px;
    max-width: 400px;
    text-align: center;
    border: thin solid $black;
    border-radius: 20px;
    transform: translate(-50%, -50%);
    background: #fff;
  }
  input {
    display: block;
    margin: 10px auto;
    font-size: 16px;
    border: thin solid $black;
    border-radius: 10px;
  }
  button {
    display: block;
    margin: 0 auto;
    font-size: 16px;
    background: $black;
    border: 0;
    color: #fff;
    padding: 5px 15px;
    border-radius: 15px;
  }
}

.container {
  display: flex;
  flex-wrap: wrap;
  height: 100vh;
  padding: 10px;
  align-items: flex-end;
  .content {
    position: relative;
    width: 100%;
    max-width: 452px;
    margin: auto;
    background: #fff;
    @extend %transition;
  }
}

.matrix {
  display: grid;
  margin: auto;
  width: 100%;
  grid-template-columns: 1fr 1fr 1fr;
  border: thin solid $black;
}

.value {
  position: relative;
  padding-top: 100%;
  border: thin solid $black;
  font-size: 100px;
  &::before {
    position: absolute;
    top: 50%;
    left: 50%;
    color: $black;
    transform: translate(-50%, -50%);
  }
  @media (min-width: 480px) {
    font-size: 130px;
  }
  @media (min-width: 768px) {
    font-size: 160px;
  }
}

.player1::before {
  content: 'X';
  color: $blue;
}

.player2::before {
  content: 'O';
  color: $green;
}

.winner-line {
  position: absolute;
  display: none;
  left: 50%;
  width: 94%;
  height: 10px;
  transform: translateX(-50%);
  background: $red;
  z-index: 1;
  &.reverse{
    display: block;
    transform: translateX(-50%) rotate(90deg);
    transform-origin: left;
    top: 3%;
    &.index0{
      left: 64%;
    }
    &.index1{
      left: 97%;
    }
    &.index2{
      left: 130%;
    }
  }
  &.diagonal{
    display: block;
    top: 7%;
    width: 120%;
    transform-origin: left;
    &.index0{
      transform: translateX(-35%) rotate(45deg);
    }
    &.index1{
      transform: translateX(35%) rotate(135deg);
    }
  }
  &.normal{
    display: block;
    &.index0{
      top: 17%;
    }
    &.index1{
      top: 50%;
    }
    &.index2{
      top: 82%;
    }
  }
}

.data-players {
  display: flex;
  width: 100%;
  text-align: center;
  flex-wrap: wrap;
  > div {
    margin: 10px auto;
    width: 50%;
    @media (min-width: 480px) {
      width: 33.3333333%;
    }
  }
  .game-count{
    order: 1;
    @media (min-width: 480px) {
      order: 0;
    }
  }
}

.restart {
  width: 100%;
  margin: 20px auto;
  text-align: center;
}

.warning {
  animation: warning-animation .2s ease-out;
}

@keyframes warning-animation {
  0%{
    transform: translateX(0);
  }
  25%{
    transform: translateX(-10%);
  }
  50%{
    transform: translateX(0%);
  }
  75%{
    transform: translateX(10%);
  }
  100%{
    transform: translateX(0);
  }
}
