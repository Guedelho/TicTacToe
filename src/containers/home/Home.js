import React, {Component} from 'react';
import {connect} from 'react-redux';

import './Home.css';

import Header from '../../components/header/Header';
import Main from '../../components/main/Main';
import Modal from "../../components/modal/Modal";

import {
  updatePlayerName,
  updatePlayerTurn,
  updatePlayerVictory,
  updateGameCount,
  updateMatrix
} from "../../modules/tictactoe";

export const verifyLine = (matrix, isReverse, isDiagonal) => {
  for (let i = 0; i < matrix.length; i++) {
    if (matrix[i].every((value) => value === 1) || matrix[i].every((value) => value === 2)) {
      return {
        winner: matrix[i][0],
        lineIndex: i,
        isReverse: isReverse,
        isDiagonal: isDiagonal
      };
    }
  }

  return false;
};

export const diagonalMatrix = (matrix) => {
  let diagonalX = [];
  let diagonalY = [];

  for (let i = 0; i < matrix.length; i++) {
    diagonalX.push(matrix[i][i]);
    diagonalY.push(matrix[i][matrix.length - i - 1]);
  }

  return [diagonalX, diagonalY];
};

export const reverseMatrix = (matrix) => matrix.map((line, i) => line.map((value, j) => matrix[j][i]));

class Home extends Component {
  constructor(){
    super();

    this.state = {
      movesCount: 0,
      degCount: 0,
      result: 0,
      winnerPlayer: ''
    }
  }

  onClickBlock = (e, i, j) => {
    let matrix = this.props.matrix;

    if (!matrix[i][j] && !this.state.result) {
      this.controlMatrixAnimation(e, this.props.playerTurn);

      matrix[i][j] = this.props.playerTurn;
      this.props.updateMatrix(matrix);

      this.setState({movesCount: this.state.movesCount + 1});

      if(this.gameController(matrix)) return;

      this.props.updatePlayerTurn(this.props.playerTurn === 1 ? 2 : 1);

    } else if(!this.state.result){
      const matrix = document.querySelector(".matrix");

      matrix.classList.add("warning");

      setTimeout(() => {
        matrix.classList.remove("warning");
      }, 300);
    }
  };

  gameController = (matrix) => {
    const result = verifyLine(matrix, false, false) || verifyLine(reverseMatrix(matrix), true, false) || verifyLine(diagonalMatrix(matrix), false, true);

    if (result) {
      this.handlerResult(result);
    } else if (this.state.movesCount === 9) {
      this.setState({result: 2});
    }

    if(result || this.state.movesCount === 9){
      this.handleEndGame();
      return true;
    }
    return false;
  };

  handlerResult = (result) => {
    const winnerLine = document.querySelector(".winner-line");
    const home = document.querySelector(".Home");

    if (result.winner === 1) {
      this.props.updatePlayerVictory("player1", this.props.playerVictory1 + 1);
    } else {
      this.props.updatePlayerVictory("player2", this.props.playerVictory2 + 1);
    }

    if (result.isReverse) {
      winnerLine.classList.add("reverse");
    } else if (result.isDiagonal) {
      winnerLine.classList.add("diagonal");
    } else {
      winnerLine.classList.add("normal");
    }

    winnerLine.classList.add("index" + result.lineIndex);
    home.classList.add("winner");
    this.setState({result: 1});
  };

  handleEndGame = () => {
    this.props.updateGameCount(this.props.gameCount + 1);
    this.props.updateMatrix(null);
  };

  controlMatrixAnimation = (e, playerTurn) => {
    e.target.classList.add("type" + playerTurn);

    this.setState({degCount: this.state.degCount + 90});
  };

  resetGameInterface = (playerColor) => {
    const winnerLine = document.querySelector(".winner-line");
    const home = document.querySelector(".Home");
    const value = [...document.querySelectorAll(".value")];

    winnerLine.className = "winner-line";
    home.className = "Home " + playerColor;
    value.map(element => element.className = "value");

    this.setState({result: 0, movesCount: 0, degCount: 0});
  };

  render() {
    const {
      playerName1,
      playerVictory1,
      playerName2,
      playerVictory2,
      updatePlayerName,
      playerTurn,
      gameCount,
      matrix
    } = this.props;

    const player = {
      name: playerTurn === 1 ? playerName1:playerName2,
      color: playerTurn === 1 ? "blue":"green"
    };

    const contentTransform = `rotate(${this.state.degCount}deg)`;

    return (
      <div className={`Home ${player.color}`}>
        <Modal title={"Jogador N1, digite seu nome:"} name={"player1"} playerName={playerName1} onChangePlayerName={updatePlayerName}/>
        <Modal title={"Jogador N2, digite seu nome:"} name={"player2"} playerName={playerName2} onChangePlayerName={updatePlayerName}/>
        <Header result={this.state.result} playerName={player.name} playerColor={player.color}/>
        <Main>
          <div className="container">
            <div className="data-players">
              <div className="data-player1">
                <div className="player-name">{playerName1} - {playerVictory1}</div>
              </div>
              <div className="game-count">
                <div>Jogo - {gameCount}</div>
              </div>
              <div className="data-player2">
                <div className="player-name">{playerName2} - {playerVictory2}</div>
              </div>
            </div>
            <div className="content" style={{transform: contentTransform}}>
              <div className="winner-line">
              </div>
              <div className="matrix">
                {matrix.map((line, i) => (
                  line.map((value, j) => (
                    <div key={i + "" + j}
                         className="value"
                         onClick={(e) => this.onClickBlock(e, i, j)}>
                    </div>
                  ))
                ))}
              </div>
            </div>
            <div className="restart">
              <div onClick={(e) => this.resetGameInterface(player.color)}>Recome√ßar!</div>
            </div>
          </div>
        </Main>
      </div>
    );
  }
}

const initMapStateToProps = state => ({
  playerName1: state.tictactoe.playerName1,
  playerVictory1: state.tictactoe.playerVictory1,
  playerName2: state.tictactoe.playerName2,
  playerVictory2: state.tictactoe.playerVictory2,
  playerTurn: state.tictactoe.playerTurn,
  gameCount: state.tictactoe.gameCount,
  matrix: state.tictactoe.matrix
});

const initMapDispatchToProps = dispatch => ({
  updatePlayerName: (e) => dispatch(updatePlayerName(e.target)),
  updatePlayerTurn: (playerTurn) => dispatch(updatePlayerTurn(playerTurn)),
  updateGameCount: (gameCount) => dispatch(updateGameCount(gameCount)),
  updatePlayerVictory: (player, victory) => dispatch(updatePlayerVictory(player, victory)),
  updateMatrix: (matrix) => dispatch(updateMatrix(matrix))
});

export default connect(initMapStateToProps, initMapDispatchToProps)(Home);
