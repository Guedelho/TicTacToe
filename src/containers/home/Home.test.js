import React from 'react';
import { shallow } from 'enzyme';
import Home, { verifyLine, diagonalMatrix, reverseMatrix } from './Home';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
 
const initialState = {
  tictactoe: {
    playerName1: '',
    playerVictory1: 0,
    playerName2: '',
    playerVictory2: 0,
    playerTurn: 1,
    gameCount: 0,
    matrix: [[0,0,0]
            ,[0,0,0]
            ,[0,0,0]]
  }
}; 

const mockStore = configureStore([thunk]);

let wrapper;
let store;

beforeEach(() => {
  store = mockStore(initialState);
  wrapper = shallow(<Home store={store}/>);
})

it('renders without crashing', () => {
  expect(wrapper).toMatchSnapshot();
});

it('testing reverseMatrixFunction', () => {
  const mockMatrix = [[1,1,1]
                  ,[0,0,0]
                  ,[0,0,0]];
  expect(reverseMatrix(mockMatrix)).toEqual([[1,0,0]
                                            ,[1,0,0]
                                            ,[1,0,0]]);
});

it('testing diagonalMatrixFunction', () => {
  const mockMatrix = [[1,0,0]
                  ,[0,1,0]
                  ,[0,0,1]];
  expect(diagonalMatrix(mockMatrix)).toEqual([[1,1,1]
                                             ,[0,1,0]]);
});

it('testing verifyLineFunction', () => {
  const mockMatrix = [[0,0,0]
                  ,[1,1,1]
                  ,[0,0,0]];
  expect(verifyLine(mockMatrix, true, false)).toEqual({
    winner: 1,
    lineIndex: 1,
    isReverse: true,
    isDiagonal: false
  });
});

