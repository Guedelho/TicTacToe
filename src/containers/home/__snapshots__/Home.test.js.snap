// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Home)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "gameCount": 0,
      "matrix": Array [
        Array [
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
        ],
      ],
      "playerName1": "",
      "playerName2": "",
      "playerTurn": 1,
      "playerVictory1": 0,
      "playerVictory2": 0,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "updateGameCount": [Function],
      "updateMatrix": [Function],
      "updatePlayerName": [Function],
      "updatePlayerTurn": [Function],
      "updatePlayerVictory": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "gameCount": 0,
        "matrix": Array [
          Array [
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
          ],
        ],
        "playerName1": "",
        "playerName2": "",
        "playerTurn": 1,
        "playerVictory1": 0,
        "playerVictory2": 0,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "updateGameCount": [Function],
        "updateMatrix": [Function],
        "updatePlayerName": [Function],
        "updatePlayerTurn": [Function],
        "updatePlayerVictory": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
